bufferserver/src/main/java/com/datatorrent/bufferserver/server/Server.java

```diff
@@ -235,81 +247,116 @@ private void handleResetRequest(ResetRequestTuple request, final AbstractLengthP
   {
-    String identifier = request.getIdentifier();
-    String type = request.getStreamType();
-    String upstream_identifier = request.getUpstreamIdentifier();
-
-    // Check if there is a logical node of this type, if not create it.
-    final LogicalNode ln;
-    if (subscriberGroups.containsKey(type)) {
-      //logger.debug("adding to exiting group = {}", subscriberGroups.get(type));
-      /*
-       * close previous connection with the same identifier which is guaranteed to be unique.
-       */
-      AbstractLengthPrependerClient previous = subscriberChannels.put(identifier, connection);
-      if (previous != null) {
-        eventloop.disconnect(previous);
-      }
-
-      ln = subscriberGroups.get(type);
+    try {
       serverHelperExecutor.submit(new Runnable()
       {
         @Override
         public void run()
         {
-          ln.boot(eventloop);
-          ln.addConnection(connection);
-          ln.catchUp();
-        }
-      });
-    } else {
-      /*
-       * if there is already a datalist registered for the type in which this client is interested,
-       * then get a iterator on the data items of that data list. If the datalist is not registered,
-       * then create one and register it. Hopefully this one would be used by future upstream nodes.
-       */
-      final DataList dl;
-      if (publisherBuffers.containsKey(upstream_identifier)) {
-        dl = publisherBuffers.get(upstream_identifier);
-        //logger.debug("old list = {}", dl);
-      } else {
-        dl = Tuple.FAST_VERSION.equals(request.getVersion()) ?
-            new FastDataList(upstream_identifier, blockSize, numberOfCacheBlocks) :
-            new DataList(upstream_identifier, blockSize, numberOfCacheBlocks);
-        publisherBuffers.put(upstream_identifier, dl);
-        //logger.debug("new list = {}", dl);
-      }
+          final String upstream_identifier = request.getUpstreamIdentifier();
 
-      long skipWindowId = (long)request.getBaseSeconds() << 32 | request.getWindowId();
-      ln = new LogicalNode(identifier, upstream_identifier, type, dl.newIterator(skipWindowId), skipWindowId);
+          /*
+           * if there is already a datalist registered for the type in which this client is interested,
+           * then get a iterator on the data items of that data list. If the datalist is not registered,
+           * then create one and register it. Hopefully this one would be used by future upstream nodes.
+           */
+          DataList dl = publisherBuffers.get(upstream_identifier);
+          if (dl == null) {
+            dl = Tuple.FAST_VERSION.equals(request.getVersion()) ?
+                new FastDataList(upstream_identifier, blockSize, numberOfCacheBlocks) :
+                new DataList(upstream_identifier, blockSize, numberOfCacheBlocks);
+            DataList odl = publisherBuffers.putIfAbsent(upstream_identifier, dl);
```

Comment:
Why putIfAbsent, is there a chance this code could be executing in separate threads? This is running within a single thread executor isn't it?